pkg/k8shandler/deployment.go

```diff
@@ -554,48 +563,63 @@ func (node *deploymentNode) isChanged() bool {
-			logrus.Debugf("Resource '%s' has different CPU limit than desired", node.self.Name)
-			nodeContainer.Resources.Limits[v1.ResourceCPU] = *desiredContainer.Resources.Limits.Cpu()
-			changed = true
-		}
-		// Check memory limits
-		if desiredContainer.Resources.Limits.Memory().Cmp(*nodeContainer.Resources.Limits.Memory()) != 0 {
-			logrus.Debugf("Resource '%s' has different Memory limit than desired", node.self.Name)
-			nodeContainer.Resources.Limits[v1.ResourceMemory] = *desiredContainer.Resources.Limits.Memory()
-			changed = true
-		}
-		// Check CPU requests
-		if desiredContainer.Resources.Requests.Cpu().Cmp(*nodeContainer.Resources.Requests.Cpu()) != 0 {
-			logrus.Debugf("Resource '%s' has different CPU Request than desired", node.self.Name)
-			nodeContainer.Resources.Requests[v1.ResourceCPU] = *desiredContainer.Resources.Requests.Cpu()
-			changed = true
+	if nodeContainer.Resources.Limits == nil {
+		nodeContainer.Resources.Limits = v1.ResourceList{}
+	}
+
+	// check that both exist
+
+	if nodeContainer.Image != desiredContainer.Image {
+		logrus.Debugf("Resource '%s' has different container image than desired", node.self.Name)
+		nodeContainer.Image = desiredContainer.Image
+		changed = true
+	}
+
+	if desiredContainer.Resources.Limits.Cpu().Cmp(*nodeContainer.Resources.Limits.Cpu()) != 0 {
+		logrus.Debugf("Resource '%s' has different CPU (%+v) limit than desired (%+v)", node.self.Name, *nodeContainer.Resources.Limits.Cpu(), desiredContainer.Resources.Limits.Cpu())
+		nodeContainer.Resources.Limits[v1.ResourceCPU] = *desiredContainer.Resources.Limits.Cpu()
+		if nodeContainer.Resources.Limits.Cpu().IsZero() {
+			delete(nodeContainer.Resources.Limits, v1.ResourceCPU)
 		}
-		// Check memory requests
-		if desiredContainer.Resources.Requests.Memory().Cmp(*nodeContainer.Resources.Requests.Memory()) != 0 {
-			logrus.Debugf("Resource '%s' has different Memory Request than desired", node.self.Name)
-			nodeContainer.Resources.Requests[v1.ResourceMemory] = *desiredContainer.Resources.Requests.Memory()
-			changed = true
+		changed = true
+	}
+	// Check memory limits
+	if desiredContainer.Resources.Limits.Memory().Cmp(*nodeContainer.Resources.Limits.Memory()) != 0 {
+		logrus.Debugf("Resource '%s' has different Memory limit than desired", node.self.Name)
+		nodeContainer.Resources.Limits[v1.ResourceMemory] = *desiredContainer.Resources.Limits.Memory()
+		if nodeContainer.Resources.Limits.Memory().IsZero() {
+			delete(nodeContainer.Resources.Limits, v1.ResourceMemory)
 		}
-
-		node.self.Spec.Template.Spec.Containers[index] = nodeContainer
+		changed = true
+	}
+	// Check CPU requests
+	if desiredContainer.Resources.Requests.Cpu().Cmp(*nodeContainer.Resources.Requests.Cpu()) != 0 {
+		logrus.Debugf("Resource '%s' has different CPU Request than desired", node.self.Name)
+		nodeContainer.Resources.Requests[v1.ResourceCPU] = *desiredContainer.Resources.Requests.Cpu()
```

Comment:
we don't need to do the `IsZero` checking for `Requests` cpu and memory?

Reply:
updated to check requests